File: Problems/0941_Valid_Mountain_Array/Valid_Mountain_Array.py
# coding: utf-8

import os
import sys
import time

class Solution:
    def validMountainArray(self, A):
        if len(A) < 3 or A[0] > A[1]:
            return False

        peak_passed = False
        for i in range(1, len(A)):
            if A[i-1] == A[i]:
                return False
            elif not peak_passed and A[i-1] > A[i]:
                peak_passed = True
            elif peak_passed and A[i-1] < A[i]:
                return False
        return peak_passed

    def validMountainArray_work(self, A):
        """
        :type A: List[int]
        :rtype: bool
        """
        direction, count = 1, 0
        for i in range(1, len(A)):
            if A[i] == A[i - 1]:
                return False
            if (A[i] - A[i - 1])*direction > 0:
                continue
            else:
                if count == 0 and i == 1:
                    return False
                elif count > 2:
                    return False
                count += 1
                direction *= -1
        if count == 1:
            return True
        else:
            return False

def str_to_int_array(flds):
    nums = [0]*len(flds)
    for i in range(len(flds)):
        nums[i] = int(flds[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("[","").replace("]","").rstrip().split(",")
    A = str_to_int_array(flds)

    time0 = time.time()

    sl = Solution()
    result = sl.validMountainArray(A)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()

File: Problems/0900_0999/0941_Valid_Mountain_Array/Project_CS/Valid_Mountain_Array.cs
using System;
using System.Linq;

public class Solution
{
    public bool ValidMountainArray(int[] A)
    {
        if (A.Length < 3 || A[0] > A[1])
            return false;

        bool peak_passed = false;
        for (int i = 1; i < A.Length; ++i)
        {
            if (A[i-1] == A[i])
                return false;
            else if (!peak_passed && A[i-1] > A[i])
                peak_passed = true;
            else if (peak_passed && A[i-1] < A[i])
                return false;
        }
        return peak_passed;        
    }

    public int[] str_to_int_array(string[] flds)
    {
        if (flds.Length <= 0)
            return null;

        int[] nums = new int[flds.Length];
        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0].ToString();
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i].ToString();
        }

        return results;
    }

    public string output_str_array(string[] flds)
    {
        string results = "";
        if (flds.Length <= 0)
            return results;
        
        results = flds[0];
        for (int i = 1; i < flds.Length; ++i)
        {
            results += ", " + flds[i];
        }

        return results;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Trim().Split(',');

        int[] A = str_to_int_array(flds);
        Console.WriteLine("A = " + output_int_array(A));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        bool results = ValidMountainArray(A);
        Console.WriteLine("result = " + results.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
